pennylane/transforms/core/transform_dispatcher.py

```diff
@@ -0,0 +1,298 @@
+
+    Convert a transform that has the signature (tape -> Sequence(tape), fn) to a transform dispatcher that can act
+    on tape, qfunc and qnode.
+
+    .. warning::
+
+        This class is developer-facing and should not be used directly.
+    """
+
+    def __init__(
+        self, transform, expand_transform=None, classical_cotransform=None
+    ):  # pylint:disable=redefined-outer-name
+        self._transform = transform
+        functools.update_wrapper(self, transform)
+        self._expand_transform = expand_transform
+        self._classical_cotransform = classical_cotransform
+
+    def __call__(self, *targs, **tkwargs):
+        obj = None
+
+        if targs:
+            # assume the first argument passed to the transform
+            # is the object we wish to transform
+            obj, *targs = targs
+
+        if isinstance(obj, qml.tape.QuantumTape):
+            return self._transform(obj, *targs, **tkwargs)
+        if isinstance(obj, qml.QNode):
+            return self._qnode_transform(
+                obj,
+                targs,
+                tkwargs,
+            )
+        if callable(obj):
+            return self._qfunc_transform(obj, targs, tkwargs)
+
+        raise TransformError(
+            "The object on which is the transform is applied is not valid. It can only be a "
+            "tape, a QNode or a qfunc."
+        )
+
+    @property
+    def transform(self):
+        """Return the quantum transform."""
+        return self._transform
+
+    @property
+    def expand_transform(self):
+        """Return the expand transform."""
+        return self._expand_transform
+
+    @property
+    def classical_cotransform(self):
+        """Return the classical co-transform."""
+        return self._classical_cotransform
+
+    def _qfunc_transform(self, qfunc, targs, tkwargs):
+        """Apply the transform on a quantum function."""
+
+        def qfunc_transformed(*args, **kwargs):
+            tape = qml.tape.make_qscript(qfunc)(*args, **kwargs)
+            transformed_tapes, _ = self._transform(tape, *targs, **tkwargs)
+
+            if len(transformed_tapes) != 1:
+                raise TransformError(
+                    "Impossible to dispatch your transform on quantum function, because more than "
+                    "one tape is returned"
+                )
+
+            transformed_tape = transformed_tapes[0]
+
+            for op in transformed_tape.circuit:
+                qml.apply(op)
+
+        return qfunc_transformed
+
+    def _qnode_transform(self, qnode, targs, tkwargs):
+        """Apply the transform on a QNode. It populates the transform program of a QNode"""
+        if self.expand_transform:
+            qnode.add_transform(TransformContainer(self._expand_transform))
+        qnode.add_transform(
+            TransformContainer(self._transform, targs, tkwargs, self._classical_cotransform)
+        )
+        return qnode
```

Comment:
I think here we want the inplace to work because we want to be able to stack multiple decorators

```
@my_transform_1
@my_transform_2
@qml.qnode(dev=device):
....
```
and also similarly do:

```
my_transform_1(my_transform_2(qnode)))(args)
```

Reply:
Yes my comment was not relevant:

The question is:

```
@qml.qnode(dev=dev):
def circuit():
```
do we want to transform the original qnode or do we need to create a new one.
```
my_transform(my_transform(circuit))
circuit(x, y) #apply both transforms
```
or 
```
circuit_transformed = my_transform(my_transform(circuit))
circuit(x, y) #apply original qnode
circuit_transformed(x, y) # apply transformed qnode
```

Reply:
@trbromley What is the best from the product perspective?